File: 프로그래머스/1/388351. 유연근무제/유연근무제.java
import java.util.*;
import java.text.*;

class Solution {
    public int solution(int[] schedules, int[][] timelogs, int startday) {
        SimpleDateFormat sdf = new SimpleDateFormat("HHmm");
        int n = schedules.length;
        boolean[] flag = new boolean[n];
        int curDay = startday;
         
       	Date[] formattedSchedules = Arrays.stream(schedules)
            .mapToObj(s -> {
                try {
                    Date date = sdf.parse(String.format("%04d", s));
                    Calendar car = Calendar.getInstance();
                    car.setTime(date);
                    car.add(Calendar.MINUTE, 10);
                    return car.getTime();
                } catch (ParseException e) {
                    throw new RuntimeException(e);
                }
            })
            .toArray(Date[]::new);
        
        for (int i=0; i<7; i++) {
            if ((curDay - 1) % 7 < 5) {
                for (int j=0; j<n; j++) {
                   	try {
                        Date curTimeLog = sdf.parse(String.format("%04d", timelogs[j][i]));
                        if (curTimeLog.after(formattedSchedules[j])) {
                            flag[j] = true;
                        }
                    } catch (ParseException e) {
                        throw new RuntimeException(e);
                    }
                }
            }
            curDay++;
        }
        
        int count = 0;
        for (int i=0; i<n; i++) {
            if (!flag[i]) {
                count++;
            }
        }
        return count;
    }
}

File: 프로그래머스/1/388351. 유연근무제/유연근무제.js
function solution(schedules, timelogs, startday) {
    const n = schedules.length;
    const flags = new Array(n).fill(true);
    
    for (let i = 0; i < n; i++) {
        let day = startday; // 요일을 개별적으로 관리
        
        for (let j = 0; j < 7; j++) { 
            // 주말이면 넘어감 (j를 증가시키지 않고 요일만 갱신)
            if (day === 6 || day === 7) {
                day = (day % 7) + 1;
                continue;
            }
            
            const timelog = timelogs[i][j];

            // 출근 기록이 없거나, 지각한 경우만 false 처리
            if (timelog === undefined || timeToMin(timelog) > timeToMin(schedules[i]) + 10) {
                flags[i] = false;
            }
            
            // 다음 요일로 이동
            day = (day % 7) + 1;
        }
    }
    
    return flags.filter(flag => flag).length;
}

function timeToMin(x) {
    return Math.floor(x / 100) * 60 + x % 100;
}
